python_modules/libraries/dagster-openai/dagster_openai/resources.py

```diff
@@ -0,0 +1,254 @@
+                i += 1
+            # Wrap the method.
+            curr.__setattr__(
+                attribute_names[i],
+                with_usage_metadata(
+                    context=context,
+                    output_name=output_name,
+                    func=curr.__getattribute__(attribute_names[i]),
+                ),
+            )
+
+    def setup_for_execution(self, context: InitResourceContext) -> None:
+        # Set up an OpenAI client based on the API key.
+        self._client = Client(api_key=self.api_key)
+
+    @contextmanager
+    def get_client(
+        self,
+        context: AssetExecutionContext,
+        output_name: Optional[str] = None,
+        api_resources: Optional[List[ApiResourceEnum]] = None,
+    ) -> Generator[Client, None, None]:
+        """Returns an ``openai.Client`` for interacting with the OpenAI API.
+
+        By default, the client comes with wrapped endpoints
+        for three API resources, Completions, Embeddings and Chat,
+        allowing to log the API usage in the asset metadata.
+
+        Note that, if required by your use case, you can specify the API resources
+        for which you would like the endpoints to be wrapped.
+        Legacy and current resources of OpenAI API v1 can be specified using `ApiResourceEnum`.
+
+        Examples:
+        .. code-block:: python
+
+            import os
+
+            from dagster import asset, define_asset_job, EnvVar
+            from dagster_openai import OpenAIResource, ApiResourceEnum
+
+
+            @asset(compute_kind="OpenAI")
+            def openai_asset(context: AssetExecutionContext, openai: OpenAIResource):
+                api_resources = [
+                    ApiResourceEnum.CHAT,
+                    ApiResourceEnum.FINE_TUNING,
+                ]
+                with openai.get_client(context, api_resources=api_resources) as client:
+                    client.chat.completions.create(
+                        model="gpt-3.5-turbo",
+                        messages=[{"role": "user", "content": "Say this is a test"}]
+                    )
+
+            openai_asset_job = define_asset_job(name="openai_asset_job", selection="openai_asset")
+
+            defs = Definitions(
+                jobs=[openai_asset_job],
+                resources={
+                    "slack": OpenAIResource(api_key=EnvVar("OPENAI_API_KEY")),
+                },
+            )
+        """
+        # If none of the OpenAI API resource types are specified,
+        # we wrap by default the methods of
+        # `openai.resources.Completions`, `openai.resources.Embeddings`
+        # and `openai.resources.chat.Completions`.
+        # This allows the usage metadata to be captured.
+        if not api_resources:
+            api_resources = [
+                ApiResourceEnum.COMPLETIONS,
+                ApiResourceEnum.CHAT,
+                ApiResourceEnum.EMBEDDINGS,
+            ]
```

Comment:
i was imagining this to be an internal implementation to start with.

[2] seems a bit heavy IMO - maybe we could stake a separate PR for this pattern. the MVP can just not allow api_resources - it's easier and better to restrict more and open it up later.

Reply:
It makes sense, I can remove the option.